lib/charms/loki_k8s/v0/loki_push_api.py

```diff
@@ -402,6 +403,109 @@ def _validate_relation_by_interface_and_direction(
         raise Exception(f"Unexpected RelationDirection: {expected_relation_role}")
 
 
+def _validate_alert_rule(rule: dict, rule_file):
+    """This method validates if an alert rule is well formed.
+
+    Args:
+        rule: A dictionary containing an alert rule definition
+        rule_file: The rule_file name
+
+    Returns:
+        Raises an AlertRuleError exception if the alert rule is not well formed
+    """
+    missing = ["alert", "expr"]
+
+    for field in missing:
+        if field not in rule.keys():
+            message = (
+                f"Alert rule '{rule_file.name}' is not well formed. Field '{field}' is missing"
+            )
+            raise AlertRuleError(message)
+
+    if rule["expr"].find("%%juju_topology%%") < 0:
+        message = (
+            f"Alert rule '{rule_file.name}' is not well formed. "
+            + "%%juju_topology%% placeholder is not present"
+        )
+        raise AlertRuleError(message)
+
+
+@dataclasses.dataclass(frozen=True)
+class JujuTopology:
+    """Dataclass for storing and formatting juju topology information."""
+
+    model: str
+    model_uuid: str
+    application: str
+    charm_name: str
+
+    @staticmethod
+    def from_charm(charm):
+        """Factory method for creating the topology dataclass from a given charm."""
+        return JujuTopology(
+            model=charm.model.name,
+            model_uuid=charm.model.uuid,
+            application=charm.model.app.name,
+            charm_name=charm.meta.name,
+        )
+
+    @staticmethod
+    def from_relation_data(data):
+        """Factory method for creating the topology dataclass from a relation data dict."""
+        return JujuTopology(
+            model=data["model"],
+            model_uuid=data["model_uuid"],
+            application=data["application"],
+            charm_name=data["charm_name"],
+        )
+
+    @property
+    def identifier(self) -> str:
+        """Format the topology information into a terse string."""
+        return f"{self.model}_{self.model_uuid}_{self.application}"
+
+    @property
+    def short_model_uuid(self):
+        """Obtain the short form of the model uuid."""
+        return self.model_uuid[:7]
+
+    @property
+    def scrape_identifier(self):
+        """Format the topology information into a scrape identifier."""
+        return "juju_{}_{}_{}".format(
+            self.model,
+            self.short_model_uuid,
+            self.application,
+        )
+
+    @property
+    def promql_labels(self) -> str:
```

Comment:
Promql?
In Loki we have [LogQL](https://grafana.com/docs/loki/latest/logql/) ;-)

Reply:
Good catch! I think we should probably rename this method both here and in prometheus_scrape: the labels are identical here and [there](https://github.com/canonical/prometheus-operator/blob/ea351276712fd54c45635181eb5296e39274bff1/lib/charms/prometheus_k8s/v0/prometheus_scrape.py#L533).

Maybe this `juju_...` concept should be named `juju_labels` instead of `promql_labels`? @mmanciop @balbirthomas

Reply:
in this lib it could be `logql_labels`

Reply:
While really similar, the duplication is incidental, meaning that the code might have different reasons to change in the future. I'd prefer us to use a descriptive, clear name over finding one that works in both projects.

Reply:
The individual keys `juju_model`, `juju_model_uuid` etc. are probably generic and not prom/log ql specific.
Do you mean that the composite syntax "key1=val1, key2=val2" is coincidentally the same between promql and logql? And that the concern of a future change in that syntax justifies having two different concept names here?

Reply:
Scrap the probably: we _want_ them to stay stable, so that we can build links and dashboards to go from a data silo to another (e.g., logs to metrics to distributed traces).

Reply:
Good. So in the context of the JujuTopology class, do we need to differentiate between prom and loki as far as the composite string of those labels is called? Any reason we shouldn't rename here and in prom to simply `ql_labels`?

Reply:
> we want them to be entirely consistent across out telemetry

Definitely, the labels themselves, or the output, should be consistent across all our data silos.

> Any reason we shouldn't rename here and in prom to simply `ql_labels`

What is considered a `ql`? Is SQL a `ql`? NoSQL? Does it add any additional meaning to prefix it with something so vague? It kind of makes sense to be specific, either using the full `ql` it is intended for, or by describing what it is (ie. `juju_topology_labels`) don't you think?

Reply:
What it really is, is `comma_separated_labels` :)